feat(api-gen): add deprecated flag to manifest, filter duplicates

jelbourn opened the pull request:
Some symbol names appear more than once in the case of function overloads; this commit filters these duplicate entries and additionally adds a deprecated flag to the results.

josephperrott commented on bazel/api-gen/manifest/index.ts:
```diff
@@ -22,10 +23,24 @@ function main() {
+  for (const collection of apiCollections) {
+    const seen = new Set<string>();
+    collection.entries = collection.entries.filter((entry) => {
+      if (seen.has(entry.name)) return false;
```
```suggestion
      if (seen.has(entry.name)) {
        return false;
      }
```

jelbourn replied:
Done

josephperrott reviewed (approved):
LGTM Overall

Just one change please

jelbourn commented:
@josephperrott the imaginary Paul in my head was looking at me with disapproving eyes for not having unit tests, so I've also added unit tests for this (which also involved refactoring the main logic out from `main.ts`). I also realized I had a bug where functions would be marked as deprecated if _any_ of the overloads were deprecated; fixed so that it only happens if they're _all_ marked deprecated.

jelbourn commented on bazel/api-gen/manifest/generate_manifest.ts:
```diff
@@ -0,0 +1,75 @@
+    collection.entries = collection.entries.filter((entry) => {
+      const lookupKey = getApiLookupKey(collection.moduleName, entry.name);
+      if (entryLookup.has(lookupKey)) {
+        entryLookup.get(lookupKey)?.push(entry);
```
Done (it doesn't really matter either way here)

jelbourn commented on bazel/api-gen/manifest/generate_manifest.ts:
```diff
@@ -0,0 +1,75 @@
+        return false;
+      }
+
+      entryLookup.set(lookupKey, []);
```
Great catch, fixed

josephperrott reviewed (approved):
LGTM
